src/blockchain/alt_block_tree.cpp

```diff
@@ -75,105 +77,200 @@ void commitPayloadsIds(BlockIndex<AltBlock>& index,
+                 index.toPrettyString());
+  VBK_ASSERT_MSG(!hasDuplicateIds(payloads.atvs),
+                 "block %s must not contain duplicate ATVs",
+                 index.toPrettyString());
+
+  // add payload ids to the block, update the payload index
+  commitPayloadsIds<VbkBlock>(index, payloads.context, storage_);
+  commitPayloadsIds<VTB>(index, payloads.vtbs, storage_);
+  commitPayloadsIds<ATV>(index, payloads.atvs, storage_);
+
+  // save payloads on disk
+  storage_.savePayloads(payloads);
+
+  // we successfully added this block payloads
+  index.setFlag(BLOCK_HAS_PAYLOADS);
+}
+
+bool AltTree::connectBlock(index_t& index, ValidationState& state) {
+  VBK_ASSERT_MSG(index.hasFlags(BLOCK_HAS_PAYLOADS),
+                 "block %s must have payloads added",
+                 index.toPrettyString());
+  VBK_ASSERT_MSG(!index.hasFlags(BLOCK_CONNECTED),
+                 "block %s is already connected",
+                 index.toPrettyString());
+  VBK_ASSERT_MSG(!index.hasFlags(BLOCK_APPLIED),
+                 "state corruption: block %s is applied",
+                 index.toPrettyString());
 
   if (getParams().isStrictAddPayloadsOrderingEnabled()) {
-    // when StrictAddPayloadsOrdering is enabled, we do not allow to execute
-    // AddPayloads to applied chain
-    VBK_ASSERT_MSG(index.pprev->hasFlags(BLOCK_HAS_PAYLOADS),
-                   "pprev does not contain BLOCK_HAS_PAYLOADS %s",
+    VBK_ASSERT_MSG(index.pprev->hasFlags(BLOCK_CONNECTED),
+                   "the previous block of block %s must be connected",
+                   index.toPrettyString());
+    VBK_ASSERT_MSG(index.allDescendantsUnconnected(),
+                   "a descendant of block %s is connected",
+                   index.toPrettyString());
+  }
+
+  // partial stateful validation
+  // FIXME: eventually we want to perform full stateful validation here,
+  // effectively find out whether setState(index) will be successful
+
+  // BUG: applyBlock will clear BLOCK_FAILED_POP flag since it does not check
+  // for duplicates
+  if (hasDuplicates<VbkBlock>(
+          index, index.getPayloadIds<VbkBlock>(), *this, state) ||
+      hasDuplicates<VTB>(index, index.getPayloadIds<VTB>(), *this, state) ||
+      hasDuplicates<ATV>(index, index.getPayloadIds<ATV>(), *this, state)) {
+    invalidateSubtree(index, BLOCK_FAILED_POP, /*do fr=*/false);
+  }
+
+  index.setFlag(BLOCK_CONNECTED);
+  tryAddTip(&index);
+
+  if (index.isValid()) {
+    onBlockConnected.emit(index);
+  } else {
+    onInvalidBlockConnected.emit(index, state);
+  };
+
+  // connect the descendants
+  for (auto* successor : index.pnext) {
+    if (successor->hasFlags(BLOCK_HAS_PAYLOADS)) {
+      ValidationState dummy;
+      connectBlock(*successor, dummy);
+    }
+  }
+
+  return index.isValid();
+}
+
+// in !StrictAddPayloadsOrdering mode, payloads can be added to any block, which
+// may trigger incorrect AltTree behavior in certain cases
```

Comment:
then we should disable !StrictAddPayloadsOrdering?

Reply:
We use StrictAddPayloadsOrdering as a stopgap for tests. Those certain cases are basically related to duplicate detection. We can add duplicate payloads to an ancestor block after adding them to a descendant and the duplicates will not be detected and handled properly